### PR TITLE
Port OP_NEWHASH

### PR DESCRIPTION
**Pull request by noahgibbs**
This is the first one I've used that involves runtime calls, so I'm converting extern C function pointers into CodePtrs and handing them to the assembler. This is, of course, not yet testable from Rust.

### CODE REVIEW

File: yjit/src/codegen.rs [hunk 1]
```diff
@@ -1775,48 +1775,50 @@ fn gen_setlocal_wc1(jit: &mut JITState, ctx: &mut Context, cb: &mut CodeBlock, o
...
     // Save the PC and SP because we are allocating
-    jit_prepare_routine_call(jit, ctx, REG0);
+    jit_prepare_routine_call(jit, ctx, cb, REG0);
 
-    if (num) {
+    if num != 0 {
         // val = rb_hash_new_with_size(num / 2);
         mov(cb, C_ARG_REGS[0], imm_opnd(num / 2));
-        call_ptr(cb, REG0, (void *)rb_hash_new_with_size);
+        let hn_code_ptr = CodePtr::from(rb_hash_new_with_size as *mut u8);
```
Comment (maximecb) [inline yjit/src/codegen.rs]: There's a `const_ptr_opnd` that you could use to simplify this a bit.

File: yjit/src/cruby.rs [hunk 2]
```diff
@@ -78,6 +78,7 @@
 //! [GhostCell]: http://plv.mpi-sws.org/rustbelt/ghostcell/
 
 use std::convert::From;
+use std::os::*;
```
Comment (maximecb) [inline yjit/src/cruby.rs]: Probably shouldn't import all of this module. You can specify specific things to import like just c_int, c_uint and such.

File: yjit/src/codegen.rs (hunk 1 above)
Comment (noahgibbs) [reply yjit/src/codegen.rs]: Would I use something other than call_ptr? const_ptr_opnd returns an X86Opnd, and I'm not sure how I'd use that.

Comment (noahgibbs) [reply yjit/src/codegen.rs]: Ah, maybe I can just call it, without the 32-bit offset stuff in call_ptr. Okay, I can do that.

File: yjit/src/cruby.rs (hunk 2 above)
Comment (noahgibbs) [reply yjit/src/cruby.rs]: Done.

File: yjit/src/codegen.rs [hunk 3]
```diff
@@ -1786,8 +1786,9 @@ fn gen_newhash(jit: &mut JITState, ctx: &mut Context, cb: &mut CodeBlock, ocb: &
     if num != 0 {
         // val = rb_hash_new_with_size(num / 2);
         mov(cb, C_ARG_REGS[0], imm_opnd(num / 2));
-        let hn_code_ptr = CodePtr::from(rb_hash_new_with_size as *mut u8);
-        call_ptr(cb, REG0, hn_code_ptr);
+        let hn_code_opnd = const_ptr_opnd(rb_hash_new_with_size as *mut u8);
+        mov(cb, REG0, hn_code_opnd);
+        call(cb, REG0);
```
Comment (maximecb) [inline yjit/src/codegen.rs:1791]: Actually call_ptr was the right thing to do. I was wrong, const_ptr_opnd wasn't appropriate here. Will fix it.

**Pull request merged**